app/hooks/search.tsx

```diff
@@ -0,0 +1,179 @@
+        }
+      }
+      prevPosition = currPosition
+    }
+
+    return score / totalWeight
+  }
+
+  return questions
+    .map(({pageid, title}) => {
+      const normalized = normalize(title)
+      return {
+        pageid,
+        title,
+        normalized,
+        model: 'plaintext',
+        score: scoringFn(normalized),
+      }
+    })
+    .sort(byScore)
+    .slice(0, numResults)
+    .filter(({score}) => score > 0)
+}
+
+/**
+ * Ignore unimportant details for similarity comparison
+ */
+const normalize = (question: string) =>
+  question
+    .toLowerCase()
+    .replace(/[^\w ]|\b(?:an?|the?)\b/g, '')
+    .replace(/(\w{2})s\b/g, '$1') // cannot use lookbehind (?<=...) because not supported on Safari
+    .replace(/\s+|_|&\s*/g, ' ')
+    .trim()
+
+const updateQueue = (item: string) => (queue: string[]) => {
+  if (item == queue[queue.length - 1]) {
+    // the lastest search is the same a this one - ignore all other searches
+    // as this is what was desired
+    return []
+  }
+  const pos = queue.indexOf(item)
+  if (pos < 0) {
+    // If the item isn't in the queue, then do nothing - it was cleared by previous results
+    return queue
+  }
+  // Remove the first occurance of the item, i.e. the oldest search. It's possible for there
+  // to be multiple such values, but they weren't the most recent searches, so who cares. They
+  // might as well be left in though, for bookkeeping reasons
+  queue.splice(pos, 1)
+  return queue
+}
+
+/**
+ * Configure the search engine.
+ *
+ * This will download the embeddings for semantic search, but until that is set up, will
+ * use baseline search over the list of questions already loaded on the site.
+ * Searches containing only one or two words will also use the baseline search
+ */
+export const useSearch = (onSiteQuestions: MutableRefObject<Question[]>, numResults = 5) => {
+  const resultsProcessor = useRef((data: any): void => {
+    data // This is here just to get typescript to stop complaining...
+  })
+  const tfWorkerRef = useRef<Worker>()
+  const [queue, setQueue] = useState([] as string[])
+  const [results, setResults] = useState([] as SearchResult[])
+
+  useEffect(() => {
+    const makeWorker = async () => {
+      const worker = new Worker('/tfWorker.js')
+      worker.addEventListener('message', ({data}) => {
+        if (data.status == 'ready') {
+          tfWorkerRef.current = worker
+        } else if (data.searchResults) {
+          resultsProcessor.current(data)
+        }
+      })
+    }
+    makeWorker()
+  }, [resultsProcessor, tfWorkerRef])
+
+  resultsProcessor.current = ({searchResults, userQuery}) => {
+    if (!searchResults) return
+
+    if (queue[queue.length - 1] == userQuery) {
+      setResults(searchResults)
+    }
+    setQueue(updateQueue(userQuery))
+  }
+
+  // Each search query gets added to the queue of searched items - the idea
+  // is that only the last item is important - all other searches can be ignored.
```

Comment:
this could be done with a simple variable to mark what the last search was, but a queue seems prettier? I'm not sure why